File: Python/Problems/Problems/problem3152.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem3152(Problem):
    def test(self) -> bool:
        nums = [3, 4, 1, 2, 6]
        queries = [[0, 4]]

        expected = [False]

        result = self.isArraySpecial(nums, queries)

        print1DArray(result)

        return expected == result

    def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
        n = len(nums)
        lastSame = [0]
        for i in range(1, n):
            if nums[i] % 2 == nums[i - 1] % 2:
                lastSame.append(i)
            else:
                lastSame.append(lastSame[i - 1])

        m = len(queries)
        result = []
        for i in range(m):
            if lastSame[queries[i][1]] <= queries[i][0]:
                result.append(True)
            else:
                result.append(False)

        return result

File: C++/Problems/Problems/problem_3152.cpp
#include "problem_3152.h"

bool problem_3152::test() {
    std::vector nums { 3, 4, 1, 2, 6 };
    std::vector<std::vector<int>> queries { { 0, 4 } };

    std::vector expected { false };

    auto result = isArraySpecial(nums, queries);

    print1DVector(result);

    return expected == result;
}

std::vector<bool> problem_3152::isArraySpecial(std::vector<int> &nums, std::vector<std::vector<int>> &queries) {
    int n = nums.size();
    std::vector<int> lastSame(n);
    lastSame[0] = 0;
    for (int i = 1; i < n; ++i)
    {
        if (nums[i] % 2 == nums[i - 1] % 2)
            lastSame[i] = i;
        else
            lastSame[i] = lastSame[i - 1];
    }

    int m = queries.size();
    std::vector result(m, false);
    for (int i = 0; i < m; ++i)
    {
        if (lastSame[queries[i][1]] <= queries[i][0])
            result[i] = true;
    }

    return result;
}
